hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/LeaseManager.java

```diff
@@ -530,35 +530,37 @@ public void setLeasePeriod(long softLimit, long hardLimit) {
     /** Check leases periodically. */
     @Override
     public void run() {
-      for(; shouldRunMonitor && fsnamesystem.isRunning(); ) {
-        boolean needSync = false;
-        try {
-          // sleep now to avoid infinite loop if an exception was thrown.
-          Thread.sleep(fsnamesystem.getLeaseRecheckIntervalMs());
-
-          // pre-filter the leases w/o the fsn lock.
-          Collection<Lease> candidates = getExpiredCandidateLeases();
-          if (candidates.isEmpty()) {
-            continue;
-          }
-
-          fsnamesystem.writeLockInterruptibly();
+      try {
+        for (; shouldRunMonitor && fsnamesystem.isRunning(); ) {
+          boolean needSync = false;
           try {
-            if (!fsnamesystem.isInSafeMode()) {
-              needSync = checkLeases(candidates);
+            // sleep now to avoid infinite loop if an exception was thrown.
+            Thread.sleep(fsnamesystem.getLeaseRecheckIntervalMs());
+
+            // pre-filter the leases w/o the fsn lock.
+            Collection<Lease> candidates = getExpiredCandidateLeases();
+            if (candidates.isEmpty()) {
+              continue;
             }
-          } finally {
-            fsnamesystem.writeUnlock("leaseManager");
-            // lease reassignments should to be sync'ed.
-            if (needSync) {
-              fsnamesystem.getEditLog().logSync();
+
+            fsnamesystem.writeLockInterruptibly();
+            try {
+              if (!fsnamesystem.isInSafeMode()) {
+                needSync = checkLeases(candidates);
+              }
+            } finally {
+              fsnamesystem.writeUnlock("leaseManager");
+              // lease reassignments should to be sync'ed.
+              if (needSync) {
+                fsnamesystem.getEditLog().logSync();
+              }
             }
+          } catch (InterruptedException ie) {
+            LOG.debug("{} is interrupted", name, ie);
           }
-        } catch(InterruptedException ie) {
-          LOG.debug("{} is interrupted", name, ie);
-        } catch(Throwable e) {
-          LOG.warn("Unexpected throwable: ", e);
         }
+      } catch (Throwable e) {
```

Comment:
If we move the exception to here, will any unexpected exception cause the lease monitor thread to exit? Will anything restart it?

The lease monitor is a critical service running inside the namenode - if it fails completely, then it should probably cause the Namenode to abort too.

If it is having problems, then it is probably good it creates some noise in the logs. Otherwise, if the Namenode does not exit too, files can fail to get closed by the lease monitor thread much later than the error occurs, which will may be hard to trace to the lease monitor error / exiting.

The example you mentioned, with a negative sleep interval - we should fix that to ensure the interval cannot get set to a value less than or equal to zero. After doing that, I am not sure if we really need to adjust the exception handler here, and allow the lease monitor to sleep and re-try as it already does.

Reply:
I'm not sure that adjusting the location of the catch exception would have any other impact. I noticed the approach of NameNodeResourceMonitor and therefore wanted to keep their style consistent. 

[NameNodeResourceMonitor ](https://github.com/apache/hadoop/blame/trunk/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java#L4505-L4530)

I think adding a check on the range of values would indeed be a more prudent approach. I will submit another PR that only checks the value range.

Apart from that, do you think the resourceRecheckInterval used within the NameNodeResourceMonitor needs to add a check for the range of values, incorrect values can cause the NameNodeResourceMonitor to shut down

Reply:
The NamenodeResoureMonitor thread would also exit if it encounters an unexpected exception, and that is probably not good. However the ResourceMonitor is less critical than the LeaseMonitor. Ideally, there should be a bound check on any configuration to prevent illegal values.

I do not have any examples, but it may be possible that some transient issue can cause some exception in the LeaseManager, and it would recover on a subsequent run, but if we keep the change here, then the thread will exit and no more runs will be attempted. In my opinion, it is better that the exception is caught, and that it retries again, logging the error each time. If it never recovers, then the namenode will need restarted after fixing the underlying issue.

If you check `NameNodeEditLogRoller` inside FSNameSystem, you can see it catches its exceptions inside the run loop and will try again on failure after the sleep interval.

I think we should revert this part of the change, but keep the new range check you added instead.

Reply:
> I think we should revert this part of the change, but keep the new range check you added instead.
I totally agree with your point of view